### PR TITLE
CO-RE basis

### PR DESCRIPTION
**Pull request by thiagoftsm**
### Description
This PR is bringing the basis for we move forward with eBPF CO-RE. I am also already converting the eBPF program `syncfs` to test the whole structure.

This current PR is assuming we will deliver for users the headers inside `co-re/includes/` and after this we will compile the code when the kernel has support for BTF.

### Test

1. Clone branch
2. Run `./tools/check-kernel-core.sh` to test your kernel. If you do not receive any warning, your host has support for BTF.
3. Install `bpftool` on your host.
4. Execute commands: `cd co-re && make`
5. Finally run: `./tests/syncfs`. If you do not have any warning, everything was fine during the tests.

### CODE REVIEW

File: co-re/Makefile [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+_LIBC ?= glibc
+
+APPS = syncfs \
+       #
+
+all: compress
+
+libbpf:
+	cd ../libbpf/src && mkdir root build && $(MAKE) BUILD_STATIC_ONLY=1 OBJDIR=build DESTDIR=.. INCLUDEDIR= LIBDIR= UAPIDIR= install \
```
Comment (UmanShahzad) [inline co-re/Makefile:23]:
```suggestion
	cd ../libbpf/src && mkdir -p root build && $(MAKE) BUILD_STATIC_ONLY=1 OBJDIR=build DESTDIR=.. INCLUDEDIR= LIBDIR= UAPIDIR= install \
```

**Review by UmanShahzad — COMMENTED** [review]
After running the build+test I'm getting a ton of output files and intermediate stuff laying around, un-gitignored:

```bash
$ gs
On branch core_basis
Your branch is up to date with 'thiago/core_basis'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
	modified:   Makefile
	modified:   ../libbpf (untracked content)

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	includes/syncfs.skel.h
	libbpf.a
	tests/syncfs
```

Will do a thorough code review itself soon.

File: co-re/Makefile (hunk 1 above)
Comment (thiagoftsm) [reply co-re/Makefile]: Fixed with https://github.com/netdata/kernel-collector/pull/266/commits/f873d73ab1054e8ec30b746fdd4c1107efcd5b5c

**Comment by thiagoftsm** [comment]
> After running the build+test I'm getting a ton of output files and intermediate stuff laying around, un-gitignored:

I changed `.gitignore` with https://github.com/netdata/kernel-collector/pull/266/commits/5bc0d323c87ab1b37573c23f58dd3c92ad04a6a4 , thanks!

**Comment by UmanShahzad** [comment]
@thiagoftsm I still get some trailing changes in the `libbpf` submodule:

```bash
$ git status
On branch core_basis
Your branch is up to date with 'thiago/core_basis'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
	modified:   ../libbpf (untracked content)

no changes added to commit (use "git add" and/or "git commit -a")
```

**Comment by UmanShahzad** [comment]
These are the changes:

```bash
$ git status
HEAD detached at aea40f7
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	bpf/
	libbpf.a
	pkgconfig/
	src/build/

nothing added to commit but untracked files present (use "git add" to track)
```

**Comment by thiagoftsm** [comment]
> @thiagoftsm I still get some trailing changes in the `libbpf` submodule:
> 

Libbpf is a submodule, to remove these erros, it will be necessary to run the following commands:

```
# rm -rf libbpf/
# git submodule update --init --recursive
```

**Comment by thiagoftsm** [comment]
> nothing added to commit but untracked files present (use "git add" to track)
> ```

Are you using the last commit ? :thinking:  I am not having more these errors.

**Comment by UmanShahzad** [comment]
@thiagoftsm Yeah I'm currently on the latest commit of this branch:

```bash
$ git log --oneline | head
5bc0d32 (HEAD -> core_basis, thiago/core_basis) core_basis: Update gitignore
f873d73 core_basis: Add arguments for mkdir
09d8d01 core_basis: Change Makefile to avoid unecessary copies and compression with more than one software
9767483 core_basis: Convert from kprobe to fentry
4249422 Add explanation for the motive we are using only glibc
422079d core_basis: Add static binary
732fb59 core_basis: Address haldolint warnings 2
964acbf core_basis: Address haldolint warnings
7a340f5 core_basis: Add missing file responsible to create problems
e61c256 core_basis: Add compilation script
```

I tried removing the submodule & reinitiailizing it, but after doing `cd co-re && make` the issue reappears:

```bash
$ rm -rf libbpf/

$ git submodule update --init --recursive
Submodule path 'libbpf': checked out 'aea40f7179433c7f8693526dff78383839c238ca'

$ make clean
rm -f *.o;
cd kernel/ && make clean;
make[1]: Entering directory '/home/mslm/dev/netdata/kernel-collector/kernel'
rm -f *.o *.ll
make[1]: Leaving directory '/home/mslm/dev/netdata/kernel-collector/kernel'
rm -f artifacts/*

$ cd co-re/

$ make
cd ../libbpf/src && mkdir -p root build && make BUILD_STATIC_ONLY=1 OBJDIR=build DESTDIR=.. INCLUDEDIR= LIBDIR= UAPIDIR= install \
&& cp build/*.a ../../co-re/	
make[1]: Entering directory '/home/mslm/dev/netdata/kernel-collector/libbpf/src'
  MKDIR    staticobjs
  CC       bpf.o
  CC       btf.o
  CC       libbpf.o
  CC       libbpf_errno.o
  CC       netlink.o
  CC       nlattr.o
  CC       str_error.o
  CC       libbpf_probes.o
  CC       bpf_prog_linfo.o
  CC       xsk.o
  CC       btf_dump.o
  CC       hashmap.o
  CC       ringbuf.o
  CC       strset.o
  CC       linker.o
  CC       gen_loader.o
  CC       relo_core.o
  AR       libbpf.a
  INSTALL  bpf.h libbpf.h btf.h libbpf_common.h libbpf_legacy.h xsk.h bpf_helpers.h bpf_helper_defs.h bpf_tracing.h bpf_endian.h bpf_core_read.h skel_internal.h
  INSTALL  libbpf.pc
  INSTALL  libbpf.a
make[1]: Leaving directory '/home/mslm/dev/netdata/kernel-collector/libbpf/src'
clang  -I../includes/ -I../libbpf/src/ -I./includes/ -I../libbpf -ggdb -O2 -target bpf -D__TARGET_ARCH_x86 -idirafter /usr/local/include -idirafter /usr/lib/llvm-12/lib/clang/12.0.0/include -idirafter /usr/include/x86_64-linux-gnu -idirafter /usr/include  -c syncfs.bpf.c -o syncfs.bpf.o
llvm-strip -g syncfs.bpf.o # strip useless DWARF info
bpftool gen skeleton syncfs.bpf.o > includes/syncfs.skel.h
clang  -Wall -ggdb -I../includes/ -I../libbpf/src/ -I./includes/ -I../libbpf -c syncfs.c -o syncfs.o
clang  -Wall -ggdb -L. syncfs.o -lelf -lz -lbpf -o tests/syncfs
tar -cf ../artifacts/netdata_ebpf-CO-RE-glibc.tar includes/*
if [ "${DEBUG:-0}" -eq 1 ]; then tar -uvf ../artifacts/netdata_ebpf-CO-RE-glibc.tar ../tools/check-kernel-core.sh; fi
xz ../artifacts/netdata_ebpf-CO-RE-glibc.tar
( cd ../artifacts; sha256sum netdata_ebpf-CO-RE-glibc.tar.xz > netdata_ebpf-CO-RE-glibc.tar.xz.sha256sum )
rm syncfs.bpf.o

$ git status
On branch core_basis
Your branch is up to date with 'thiago/core_basis'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
	modified:   ../libbpf (untracked content)

no changes added to commit (use "git add" and/or "git commit -a")
```

**Comment by UmanShahzad** [comment]
It's not a huge deal but I think we require that "making" artifacts from libbpf causes it to output those artifacts to our own repo, so that our `make clean` can get rid of them and our `.gitignore` applies to them.

Either that or we need to have `make clean` go into the submodule and also clean things.

Or we can just live with it :smile:

**Comment by thiagoftsm** [comment]
> Or we can just live with it smile

Now I think it is fixed with https://github.com/netdata/kernel-collector/pull/266/commits/cef080652ea94390093c1cb606e9f46f46e9cdee fixes the problem. We were missing one line inside `.gitmodules`.

**Review by UmanShahzad — APPROVED** [review]
Tests pass!

**Review by vlvkobal — APPROVED** [review]

**Review by iigorkarpov — APPROVED** [review]

File: .gitmodules [hunk 2]
```diff
@@ -1,3 +1,4 @@
 [submodule "libbpf"]
 	path = libbpf
 	url = https://github.com/netdata/libbpf
+        ignore = dirty
```
[the remark this answers is not in the document]
Comment (thiagoftsm) [reply .gitmodules:4]: The first commit of next PR will address this. Thanks!

**Pull request merged**